src/hotspot/share/prims/jvmtiTagMap.cpp

```diff
@@ -3298,117 +2975,49 @@ void JvmtiTagMap::follow_references(jint heap_filter,
-  int freed = 0;
-  int moved = 0;
-
-  JvmtiTagHashmap* hashmap = this->hashmap();
-
-  // reenable sizing (if disabled)
-  hashmap->set_resizing_enabled(true);
-
-  // if the hashmap is empty then we can skip it
-  if (hashmap->_entry_count == 0) {
-    return;
-  }
-
-  // now iterate through each entry in the table
-
-  JvmtiTagHashmapEntry** table = hashmap->table();
-  int size = hashmap->size();
-
-  JvmtiTagHashmapEntry* delayed_add = NULL;
-
-  for (int pos = 0; pos < size; ++pos) {
-    JvmtiTagHashmapEntry* entry = table[pos];
-    JvmtiTagHashmapEntry* prev = NULL;
-
-    while (entry != NULL) {
-      JvmtiTagHashmapEntry* next = entry->next();
-
-      // has object been GC'ed
-      if (!is_alive->do_object_b(entry->object_raw())) {
-        // grab the tag
-        jlong tag = entry->tag();
-        guarantee(tag != 0, "checking");
-
-        // remove GC'ed entry from hashmap and return the
-        // entry to the free list
-        hashmap->remove(prev, pos, entry);
-        destroy_entry(entry);
-
-        // post the event to the profiler
-        if (post_object_free) {
-          JvmtiExport::post_object_free(env(), tag);
+  bool is_vm_thread = Thread::current()->is_VM_thread();
+  if (!is_vm_thread) {
+    if (num_dead_entries != 0) {
+      JvmtiEnvIterator it;
+      for (JvmtiEnv* env = it.first(); env != NULL; env = it.next(env)) {
+        JvmtiTagMap* tag_map = env->tag_map_acquire();
+        if (tag_map != NULL) {
+          // Lock each hashmap from concurrent posting and cleaning
+          tag_map->unlink_and_post_locked();
         }
-
-        ++freed;
-      } else {
-        f->do_oop(entry->object_addr());
-        oop new_oop = entry->object_raw();
-
-        // if the object has moved then re-hash it and move its
-        // entry to its new location.
-        unsigned int new_pos = JvmtiTagHashmap::hash(new_oop, size);
-        if (new_pos != (unsigned int)pos) {
-          if (prev == NULL) {
-            table[pos] = next;
-          } else {
-            prev->set_next(next);
-          }
-          if (new_pos < (unsigned int)pos) {
-            entry->set_next(table[new_pos]);
-            table[new_pos] = entry;
-          } else {
-            // Delay adding this entry to it's new position as we'd end up
-            // hitting it again during this iteration.
-            entry->set_next(delayed_add);
-            delayed_add = entry;
-          }
-          moved++;
-        } else {
-          // object didn't move
-          prev = entry;
+      }
+      // there's another callback for needs_rehashing
+    }
+  } else {
+    assert(SafepointSynchronize::is_at_safepoint(), "must be");
+    JvmtiEnvIterator it;
+    for (JvmtiEnv* env = it.first(); env != NULL; env = it.next(env)) {
+      JvmtiTagMap* tag_map = env->tag_map_acquire();
+      if (tag_map != NULL && !tag_map->is_empty()) {
+        if (num_dead_entries != 0) {
+          tag_map->hashmap()->unlink_and_post(tag_map->env());
```

Comment:
The JVMTI code expects the posting to be done quite eagerly presumably during GC, before it has a chance to disable the event or some other such operation.  So the posting is done during the notification because it's as soon as possible.  Deferring to the ServiceThread had two problems.  1. the event came later than the caller is expecting it, and in at least one test the event was disabled before posting, and 2. there's a comment in the code why we can't post events with a JavaThread.  We'd have to transition into native while holding a no safepoint lock (or else deadlock).  The point of making this change was so that the JVMTI table does not need GC code to serially process the table.

Reply:
I know of nothing that leads to "presumably during GC" being a requirement. Having all pending events of some type occur before that type of event is disabled seems like a reasonable requirement, but just means that event disabling also requires the table to be "up to date", in the sense that any GC-cleared entries need to be dealt with. That can be handled just like other operations that use the table contents, rather than during the GC.  That is, use post_dead_object_on_vm_thread if there are or might be any pending dead objects, before disabling the event.

Reply:
The point of this change was not necessarily to be lazy about updating the tagmap, until someone uses it. The point was to get rid of the last annoying serial GC phase. Doing it all lazily would certainly also achieve that. But it would also lead to situations where no event is ever posted from GC to GC. So you would get the event 20 GCs later, which might come as a surprise. It did come as a surprise to some tests, so it is reasonable to assume it would come as a surprise to users too. And I don't think we want such surprises unless we couldn't deal with them. And we can.

Reply:
Kim's change to post the events from the service thread or before other JVMTI operations removes posting events from  the gc_notification, which was the objection.